File: apollo/src/main/java/io/muun/apollo/domain/errors/MessagePermissionsError.java
package io.muun.apollo.domain.errors;

import io.muun.common.api.messages.MessageSpec;
import io.muun.common.model.SessionStatus;

public class MessagePermissionsError extends RuntimeException {

    /**
     * Rich constructor.
     */
    public MessagePermissionsError(String sessionUuid,
                                   long messageId,
                                   SessionStatus currentStatus,
                                   MessageSpec spec) {

        super(String.format(
                "Sender %s sent message %s of type %s expecting %s, but session was %s",
                sessionUuid,
                Long.toString(messageId),
                spec.messageType,
                spec.permission.name(),
                currentStatus != null ? currentStatus.name() : "null"
        ));
    }
}

File: android/apollo/src/main/java/io/muun/apollo/domain/errors/notifications/MessagePermissionsError.kt
package io.muun.apollo.domain.errors.notifications

import io.muun.apollo.domain.errors.MuunError
import io.muun.common.api.messages.MessageSpec
import io.muun.common.model.SessionStatus

class MessagePermissionsError(
    sessionUuid: String,
    messageId: Long,
    currentStatus: SessionStatus?,
    spec: MessageSpec,
) : MuunError("Received a message without the right permissions") {

    init {
        metadata["sessionUuid"] = sessionUuid
        metadata["sessionStatus"] = currentStatus?.name ?: "null"
        metadata["messageId"] = messageId
        metadata["messageType"] = spec.messageType
        metadata["allowedSessionStatus"] = spec.permission.name
    }
}